model/feature/flag.go

```diff
@@ -0,0 +1,270 @@
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/cozy/cozy-stack/model/instance"
+	"github.com/cozy/cozy-stack/pkg/config/config"
+	"github.com/cozy/cozy-stack/pkg/consts"
+	"github.com/cozy/cozy-stack/pkg/couchdb"
+)
+
+type Flags struct {
+	DocID   string
+	DocRev  string
+	M       map[string]interface{}
+	Sources []*Flags
+}
+
+func (f *Flags) ID() string        { return f.DocID }
+func (f *Flags) Rev() string       { return f.DocRev }
+func (f *Flags) DocType() string   { return consts.Settings }
+func (f *Flags) SetID(id string)   { f.DocID = id }
+func (f *Flags) SetRev(rev string) { f.DocRev = rev }
+func (f *Flags) Clone() couchdb.Doc {
+	clone := Flags{DocID: f.DocID, DocRev: f.DocRev}
+	clone.M = make(map[string]interface{})
+	for k, v := range f.M {
+		clone.M[k] = v
+	}
+	return &clone
+}
+func (f *Flags) MarshalJSON() ([]byte, error) {
+	return json.Marshal(f.M)
+}
+
+func (f *Flags) UnmarshalJSON(bytes []byte) error {
+	err := json.Unmarshal(bytes, &f.M)
+	if err != nil {
+		return err
+	}
+	if id, ok := f.M["_id"].(string); ok {
+		f.SetID(id)
+		delete(f.M, "_id")
+	}
+	if rev, ok := f.M["_rev"].(string); ok {
+		f.SetRev(rev)
+		delete(f.M, "_rev")
+	}
+	return nil
+}
+
+func GetFlags(inst *instance.Instance) (*Flags, error) {
+	sources := make([]*Flags, 0)
+	m := make(map[string]interface{})
+	flags := &Flags{
+		DocID:   consts.FlagsSettingsID,
+		M:       m,
+		Sources: sources,
+	}
+	flags.addInstanceFlags(inst)
+	if err := flags.addManager(inst); err != nil {
+		return nil, err
+	}
+	if err := flags.addConfig(inst); err != nil {
+		return nil, err
+	}
+	if err := flags.addContext(inst); err != nil {
+		return nil, err
+	}
+	if err := flags.addDefaults(inst); err != nil {
+		return nil, err
+	}
+	return flags, nil
+}
+
+func (f *Flags) addInstanceFlags(inst *instance.Instance) {
+	if len(inst.FeatureFlags) == 0 {
+		return
+	}
+	m := make(map[string]interface{})
+	for k, v := range inst.FeatureFlags {
+		m[k] = v
+	}
+	flags := &Flags{
+		DocID: consts.InstanceFlagsSettingsID,
+		M:     m,
+	}
+	f.Sources = append(f.Sources, flags)
+	for k, v := range flags.M {
+		if _, ok := f.M[k]; !ok {
+			f.M[k] = v
+		}
+	}
+}
+
+func (f *Flags) addManager(inst *instance.Instance) error {
+	if len(inst.FeatureSets) == 0 {
+		return nil
+	}
+	m, err := getFlagsFromManager(inst)
+	if err != nil || len(m) == 0 {
+		return err
+	}
+	flags := &Flags{
+		DocID: consts.ManagerFlagsSettingsID,
+		M:     m,
+	}
+	f.Sources = append(f.Sources, flags)
+	for k, v := range flags.M {
+		if _, ok := f.M[k]; !ok {
+			f.M[k] = v
+		}
+	}
+	return nil
+}
+
+var (
+	cacheDuration      = 12 * time.Hour
+	managerHTTPClient  = &http.Client{Timeout: 5 * time.Second}
+	errInvalidResponse = errors.New("Invalid response from the manager")
+)
+
+func getFlagsFromManager(inst *instance.Instance) (map[string]interface{}, error) {
+	cache := config.GetConfig().CacheStorage
+	cacheKey := fmt.Sprintf("flags:%s:%v", inst.ContextName, inst.FeatureSets)
```

Comment:
I don't think we need a bijection between offers/plans and stack context, only that a context always give the same feature flags. That said, I also think it would be a good thing to have this bijection.

Reply:
There is no bijection at the moment, we need in theory the full tuple partner/offer/context/plans to got the flags. In practice, features are only on plans, associated to partner.
So you need to ask flags with lists of instance plan ids, which is provided by the cloudery too on creation and subscription.

Reply:
ok, we talked about it, feature set list is sufficient and there is not problem as feature sets will be ids of feature sets unique to the whole environement.
so no problem here (and cloudery can use or ignore context, whatever)